### ISSUE TITLE
66% of training tokens is padding

### ISSUE THREAD

**Issue by sdi1100041** [sandyresearch/parcae#8]
> **TL;DR:** the best-fit packer is fed `micro_batch_size` documents per call but
> must fill `micro_batch_size` full rows, so on FineWeb-Edu
> (`SandyResearch/fineweb-edu-shuffled`) **~67% of every batch is padding**
> (measured real-token fraction 0.337). Padding is masked from the loss, so a run
> configured for *B* tokens trains on only **~*B*/3 real tokens** — a nominal
> 80 tokens/param is effectively **~27**. We reproduced your GPT-370M / FineWeb-Edu
> baseline (CORE **17.18** vs your **17.46**, Val PPL **15.20** vs **15.79**,
> Lambada PPL **38.50** vs **40.77** — matching on all three metrics), then turned
> on packing with everything else held identical (same model, steps, wall-clock):
> **CORE 17.18 → 19.74**, **Val PPL 15.20 → 14.84**, and **Lambada PPL 38.50 → 28.14**,
> all at identical compute. The effect is independent of `micro_batch_size` and
> of model size.

## Summary

During pretraining, the best-fit sequence packer is fed only `micro_batch_size`
documents per collate call while being asked to fill `micro_batch_size` rows of
`block_size` tokens each. When the mean document is shorter than `block_size`
(the usual case for web text), supply is far below demand, the document buffer
stays drained, and rows are pad-filled. The `buffer_size` parameter that looks
intended to prevent this is stored but never used.

Net effect: a large fraction of every training batch is padding. Padding tokens
are excluded from the loss, so training is not *incorrect*, but a big share of
the compute is spent on pad positions and the effective token budget is a
fraction of the nominal one.

## Root cause (refs against `main`)

1. `scripts/train.py:262` — the training dataloader is created with
   `batch_size=cfg.micro_batch_size`, so `BestFitPackingCollator.__call__`
   receives exactly `micro_batch_size` documents per call.
2. `recpre/data_loading_utils.py` — `__call__` builds a `row_buffer` of shape
   `(len(batch), row_capacity)` = `(micro_batch_size, block_size)`, i.e. it must
   emit `micro_batch_size` fully packed rows from those same `micro_batch_size`
   documents (plus whatever little is left in `self.doc_buffer`, which is only
   ever fed by previous calls at the same starved rate).
3. `recpre/data_loading_utils.py:21` — `self.buffer_size = buffer_size`
   (`cfg.pack_buffer_size`, default **1000**, passed at `train.py:1042`) is never
   referenced inside `__call__`. The buffer is never refilled toward
   `buffer_size`, so the mechanism that would decouple docs-read from rows-emitted
   is absent.

Because documents-per-call equals rows-per-call, the padding fraction is
approximately `1 - mean_doc_len / block_size` **and is independent of
`micro_batch_size`** (raising the micro-batch raises supply and demand equally).

## Evidence

Measured with the released tokenizer (`SandyResearch/parcae-tokenizer`) on
`SandyResearch/fineweb-edu-shuffled` `sample-100BT` (i.e. **100% FineWeb-Edu,
the released training corpus** — not a custom mixture):

- document length: mean ≈ **977 tokens**, median **610**, p90 1868, long tail
  (max 25k), vs `block_size` 2048
- **0.000** of rows are ≈ one block long and only 8.5% exceed `block_size` — i.e.
  the data is raw **per-document**, not pre-packed sequences (a pre-packed
  corpus would need no packer at all)
- driving the actual `BestFitPackingCollator` with `micro_batch_size` documents
  per call (exactly as `create_dataloaders` does), the **measured real-token
  fraction is 0.337** → **~67% padding**. (The naive supply/demand ratio
  977/2048 ≈ 0.48 is only an upper bound; best-fit fragmentation — a row ends
  with fewer free slots than the shortest buffered doc — pushes it lower.)

So a nominal *N*-token run on FineWeb-Edu trains on **~*N*/3 real tokens**; a
configured 80 tokens/param budget is effectively **~27**.

### Direct impact on the paper's own metrics (reproduce, then fix)

We trained a GPT-370M on FineWeb-Edu with your pipeline/config and evaluated on
your three reported metrics. This separates "faithful reproduction" from "the
fix": the padded run should match the paper (it does, on all three), and the
only change for the packed run is removing the padding — same model, same steps,
same wall-clock.

| GPT-370M / FineWeb-Edu | CORE ↑ | Val PPL ↓ | Lambada PPL ↓ |
|---|---|---|---|
| Parcae paper (reported) | 17.46 | 15.79 | 40.77 |
| Ours, current pipeline (padded) | 17.18 | 15.20 | 38.50 |
| Ours, concatenation packing (~0 pad) | **19.74** | **14.84** | **28.14** |

The padded run reproduces all three published numbers within noise, which
indicates our setup faithfully matches yours — and, since a pre-packed corpus
would not exhibit the padding, that the released runs used this raw per-document
`pqds-pure` / `keys: ["text"]` path. Turning packing on raises **CORE by +2.56**
and cuts **Lambada PPL 38.50 → 28.14**, at identical compute. We see
same-direction, larger effects on 5-shot math (≈2× exact-match on a
math-containing corpus) and reasoning-primitive tasks.

## Minimal reproduction

```python
# tokenize a few thousand fineweb-edu-shuffled docs with the parcae tokenizer,
# then drive the collator exactly as create_dataloaders does (batch_size = micro_batch):
from recpre.data_loading_utils import BestFitPackingCollator
coll = BestFitPackingCollator(tokenizer=tok, block_size=block_size+1, buffer_size=1000)
frac = []
for i in range(0, len(docs), micro_batch):          # micro_batch docs per call
    x, y, _ = coll(docs[i:i+micro_batch])
    frac.append((x != (tok.pad_id or 0)).float().mean().item())
print(sum(frac)/len(frac))                            # ≈ 0.33 real tokens
```

## Suggested directions (maintainers' call)

The fix is to decouple *documents read* from *rows emitted*. Options, roughly in
increasing invasiveness:

1. **Honor `buffer_size`**: have the collator keep `self.doc_buffer` topped up to
   ~`buffer_size` documents and emit a fixed `micro_batch_size` rows, carrying the
   surplus across calls (requires feeding more docs per call, e.g. a larger
   dataloader `batch_size`, with the buffer as the shock absorber).
2. **Pack inside the `IterableDataset`**: have `ParquetStreamPure` (or a thin
   wrapper) concatenate tokenized docs and yield `block_size+1` sequences
   directly; the collate then just stacks. This gives ~zero padding and is the
   standard GPT-style approach; note it packs greedily across document
   boundaries (consistent with the current setup, where
   `get_attention_mask` returns `None` so attention is already cross-document).

Either way it may be worth gating behind the existing `pack_sequences` flag so
strict reproductions of published runs remain available.

## Impact

For sub-`block_size` corpora, fixing this recovers `1 / real_token_fraction`
more real tokens per unit compute — **≈3× for FineWeb-Edu** at `block_size` 2048
(measured real fraction 0.337) — at identical wall-clock, since the tensor
shapes and FLOPs per step are unchanged.

Happy to share the measurement script or open a PR if useful.

**Comment by Hprairie** [comment]
Hey. Sorry for the delay. I have looked it over and reconfirmed the issue locally; quite a shame :(. 

Honoring `buffer_size` (i.e., 1) is more what I intended to follow, as it comes from `https://github.com/karpathy/nanochat/blob/master/nanochat/dataloader.py`.

I will get this fixed asap, thanks for pointing it out.

**Comment by sdi1100041** [comment]
Hello, you can also see a dataset related issue here #10
